// @ng-web-apis/geolocation/fesm2015/ng-web-apis-geolocation.js
import * as i0 from '@angular/core';
import { InjectionToken, inject, Injectable, Inject } from '@angular/core';
import { Observable } from 'rxjs';
import { finalize, shareReplay } from 'rxjs/operators';
import { NAVIGATOR } from '@ng-web-apis/common';

const GEOLOCATION = new InjectionToken('An abstraction over window.navigator.geolocation object', {
    factory: () => inject(NAVIGATOR).geolocation,
});

const POSITION_OPTIONS = new InjectionToken('Token for an additional position options', { factory: () => ({}) });

const GEOLOCATION_SUPPORT = new InjectionToken('Is Geolocation API supported?', {
    factory: () => {
        return !!inject(GEOLOCATION);
    },
});

// TODO: Replace type with GeolocationPosition after bumping TS to 4.1.3+
// @dynamic
class GeolocationService extends Observable {
    constructor(geolocationRef, geolocationSupported, positionOptions) {
        let watchPositionId = 0;
        super(subscriber => {
            if (!geolocationSupported) {
                subscriber.error('Geolocation is not supported in your browser');
            }
            watchPositionId = geolocationRef.watchPosition(position => subscriber.next(position), positionError => subscriber.error(positionError), positionOptions);
        });
        return this.pipe(finalize(() => geolocationRef.clearWatch(watchPositionId)), shareReplay({ bufferSize: 1, refCount: true }));
    }
}
GeolocationService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: GeolocationService, deps: [{ token: GEOLOCATION }, { token: GEOLOCATION_SUPPORT }, { token: POSITION_OPTIONS }], target: i0.ɵɵFactoryTarget.Injectable });
GeolocationService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: GeolocationService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: GeolocationService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: function () { return [{ type: Geolocation, decorators: [{
                    type: Inject,
                    args: [GEOLOCATION]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [GEOLOCATION_SUPPORT]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [POSITION_OPTIONS]
                }] }]; } });

/**
 * Public API Surface of geolocation
 */

/**
 * Generated bundle index. Do not edit.
 */

export { GEOLOCATION, GEOLOCATION_SUPPORT, GeolocationService, POSITION_OPTIONS };
